ui/src/utils/types.ts

```diff
@@ -20,47 +20,63 @@ interface AsyncTaskPending {
+  data: Data;
   status: "reloading";
 }
 
-interface AsyncTaskSuccessful<D> {
-  data: D;
+interface AsyncTaskSuccessful<Data> {
+  data: Data;
   status: "successful";
 }
 
-export function isAsyncTaskDataAvailable<D, E>(
-  task: AsyncTask<D, E>
-): task is AsyncTaskReloading<D> | AsyncTaskSuccessful<D> {
+export function isAsyncTaskDataAvailable<Data, Error>(
+  task: AsyncTask<Data, Error>
+): task is AsyncTaskReloading<Data> | AsyncTaskSuccessful<Data> {
   return task.status === "reloading" || task.status === "successful";
 }
 
-export function isAsyncTaskStarting<D, E>(
-  task: AsyncTask<D, E>
+export function isAsyncTaskStarting<Data, Error>(
+  task: AsyncTask<Data, Error>
 ): task is AsyncTaskLoading | AsyncTaskPending {
   return task.status === "loading" || task.status === "pending";
 }
 
-export function hasAsyncTaskFailed<D, E>(task: AsyncTask<D, E>): task is AsyncTaskFailed<E> {
+export function hasAsyncTaskFailed<Data, Error>(
+  task: AsyncTask<Data, Error>
+): task is AsyncTaskFailed<Error> {
   return task.status === "failed";
 }
 
+// The following was implemented due to a perceived limitation of Zod:
+// https://github.com/colinhacks/zod/issues/652
+
+// Asserts whether 2 types are exactly the same.
 type Exact<T, U> = [T, U] extends [U, T] ? true : false;
 
-export const StrictSchema: <Input, Output = Input>() => <ParserOutput, ParserInput = ParserOutput>(
+// A function that accepts two types then generates a function that accepts a parser for them if the types match,
+// otherwise it produces a static error.
+export function getStrictParser<Input, Output = Input>(): <
+  ParserOutput,
+  ParserInput = ParserOutput
+>(
+  // 1. If the parser's input is exactly the same as the output of the provided zod schema,
   parser: Exact<
     ZodSchema<Output, ZodTypeDef, Input>,
     ZodSchema<ParserOutput, ZodTypeDef, ParserInput>
   > extends true
-    ? Exact<Required<Output>, Required<ParserOutput>> extends true
-      ? Exact<Required<Input>, Required<ParserInput>> extends true
-        ? ZodSchema<ParserOutput, ZodTypeDef, ParserInput>
-        : never
-      : never
-    : never
-) => ZodSchema<Output, ZodTypeDef, Input> =
-  () =>
-  <Output, Input = Output>(parser: unknown) =>
-    // eslint-disable-next-line no-type-assertion/no-type-assertion
-    parser as ZodSchema<Output, ZodTypeDef, Input>;
+    ? // 2. then if the output types match exactly,
+      Exact<Required<Output>, Required<ParserOutput>> extends true
+      ? // 3. then if the input types match exactly,
+        Exact<Required<Input>, Required<ParserInput>> extends true
+        ? // 3. return type as a ZodSchema parser (it is certain that I/O and parser I/O are identical).
+          ZodSchema<ParserOutput, ZodTypeDef, ParserInput>
+        : // 3. fail
+          never
+      : // 2. fail
+        never
+    : // 1. fail
+      never
+) => ZodSchema<Output, ZodTypeDef, Input> {
+  return (parser: ZodSchema) => parser;
```

Comment:
Nice trick typing the parser just as `ZodSchema` so we can remove the type assertion! Per my tests, this works as well as before. I've tested all the cases from the [original issue](https://github.com/colinhacks/zod/issues/652):

* Omit in a ZodSchema an optional prop of the TS type.
* Forget to make optional in a ZodSchema and optional prop of the TS type.
* Omit in a ZodSchema a required prop of the TS type.
* Make optional in a ZodSchema a required prop of the TS type.
* Add to a ZodSchema a prop that does not exist in the TS type.

Reply:
By the way, I yesterday commented that Colin McDonnell published a utility similar to our `getStrictParser`. It's named [toZod](https://github.com/colinhacks/tozod) and sadly, it's not maintained and therefore it's incompatible with the last versions of Zod.

I also commented that it uses [a slightly different version](https://github.com/colinhacks/tozod/blob/master/src/index.ts#L6) of our `Exact` utility type, in case you prefer it:

```typescript
type Exact<X, Y> = [X] extends [Y] ? ([Y] extends [X] ? true : false) : false;
```

Reply:
Yeah that doesn't make it clearer IMO. At least the current one is shorter.